### PR TITLE
✨ Validation Reports

### LINKED ISSUE #480: Implement and test validation report builder
Each type of test will output a standard machine readable output, defined in #473. The outputs should be collected and turned into a human readable markdown validation report (initial example is [here](https://github.com/d3b-center/clinical-data-flow/tree/natasha-validation/validation#sample-validation-report)) 

Additional requirements:

1. Any relationship test that fails because the "to" entity didn't exist as a column in any of the data files should be reported in a single message like this: `P1 linked to 0 <entity type>` rather than a ton of individual messages:

For example:

Good:
image

Bad:
image

2. Any test fails for more than `<max limit here>` entities should _probably_ report a summary error message and also point to an additional file that contains all of the errors:

For example: `1856 of 2208 total Participants failed this test. Output is too large to display, please see specimen-1-participant.tsv for list of all errors`

3. We should display the unique counts for all of the ID entity types up at the top of the validation report since right now we don't really know if expected counts will be given to us and if we will implement/run count tests:

image

 4. May be a good idea to also display the files that were evaluated in the validation run at the bottom of the report

image

### PR DESCRIPTION
**Pull request by znatty22**
Closes #480 

Currently outputs two types of validation reports: TSV and markdown. 

**TSV**
This isn't really a report, but 3 tables: `type_counts.tsv`,`files_validated.tsv`, `validation_results.tsv`. The `validation_results.tsv` table is a less nested, less verbose version of the original validation results coming from the graph validator. 

I added this because we were told the analysts want a tabular version of the validation results, but I think new markdown report will satisfy their needs. So we can get rid of this if we don't think its useful

**Markdown**
The markdown report is meant to be a more human friendly/readable form of the original validation results from the graph validator. The report currently looks like this:

# 📓 Data Validation Report
## 📂 Files Validated
**/Users/singhn4/Projects/pull-request-reviews/kf-lib-data-ingest/tests/data/validation/DATASET1**

| Files    |
|:---------|
| pf.csv   |
| pg.csv   |
| sfp2.csv |
| sg.csv   |
| sp.csv   |
| spf.csv  |

##  #️⃣  Counts
| Entity                              |   Count |
|:------------------------------------|--------:|
| BIOSPECIMEN.ID                      |      12 |
| PARTICIPANT.ID                      |      10 |
| FAMILY.ID                           |       7 |
| GENOMIC_FILE.URL_LIST               |       5 |
| BIOSPECIMEN.SHIPMENT_ORIGIN         |       0 |
| SEQUENCING.PLATFORM                 |       0 |
| SEQUENCING.PAIRED_END               |       0 |
| SEQUENCING.STRATEGY                 |       0 |
| SEQUENCING.LIBRARY_NAME             |       0 |
| GENOMIC_FILE.REFERENCE_GENOME       |       0 |
| GENOMIC_FILE.HARMONIZED             |       0 |
| SEQUENCING.ID                       |       0 |
| BIOSPECIMEN.SAMPLE_PROCUREMENT      |       0 |
| BIOSPECIMEN.VOLUME_UL               |       0 |
| BIOSPECIMEN.CONCENTRATION_MG_PER_ML |       0 |
| BIOSPECIMEN.SHIPMENT_DATE           |       0 |
| BIOSPECIMEN.SPATIAL_DESCRIPTOR      |       0 |
| BIOSPECIMEN.EVENT_AGE_DAYS          |       0 |
| BIOSPECIMEN.TUMOR_DESCRIPTOR        |       0 |
| BIOSPECIMEN.ANATOMY_SITE            |       0 |
| BIOSPECIMEN.TISSUE_TYPE             |       0 |
| BIOSPECIMEN.COMPOSITION             |       0 |
| BIOSPECIMEN.ANALYTE                 |       0 |
| GENOMIC_FILE.ID                     |       0 |
| PARTICIPANT.ENROLLMENT_AGE_DAYS     |       0 |
| PARTICIPANT.SPECIES                 |       0 |
| PARTICIPANT.RACE                    |       0 |
| PARTICIPANT.ETHNICITY               |       0 |
| PARTICIPANT.GENDER                  |       0 |
| SEQUENCING.INSTRUMENT               |       0 |

#### ❌ Each FAMILY.ID links to at least 1 PARTICIPANT.ID

| Errors                                           |
|:-------------------------------------------------|
| `FAMILY.ID.F12` is linked to  0 `PARTICIPANT.ID` |

| Location   | Values                                                       |
|:-----------|:-------------------------------------------------------------|
| pf.csv     | This file contains all errors in the **Errors** table above. |

#### ❌ Each BIOSPECIMEN.ID links to exactly 1 PARTICIPANT.ID

| Errors                                               |
|:-----------------------------------------------------|
| `BIOSPECIMEN.ID.S2` is linked to  ['`P2`', '`P1`']   |
| `BIOSPECIMEN.ID.S8` is linked to  0 `PARTICIPANT.ID` |

| Location   | Values                                                       |
|:-----------|:-------------------------------------------------------------|
| sfp2.csv   | This file contains all errors in the **Errors** table above. |
| spf.csv    | `P1`,`S8`,``S2``                                             |

#### ❌ Each PARTICIPANT.ID links to at least 1 BIOSPECIMEN.ID

| Errors                                                |
|:------------------------------------------------------|
| `PARTICIPANT.ID.P13` is linked to  0 `BIOSPECIMEN.ID` |
| `PARTICIPANT.ID.P11` is linked to  0 `BIOSPECIMEN.ID` |

| Location   | Values                                                       |
|:-----------|:-------------------------------------------------------------|
| pf.csv     | This file contains all errors in the **Errors** table above. |

#### ❌ Each BIOSPECIMEN.ID links to at least 1 GENOMIC_FILE.ID

| Errors                                                 |
|:-------------------------------------------------------|
| `BIOSPECIMEN.ID.NA` is linked to  0 `GENOMIC_FILE.ID`  |
| `BIOSPECIMEN.ID.S7` is linked to  0 `GENOMIC_FILE.ID`  |
| `BIOSPECIMEN.ID.S2` is linked to  0 `GENOMIC_FILE.ID`  |
| `BIOSPECIMEN.ID.S4` is linked to  0 `GENOMIC_FILE.ID`  |
| `BIOSPECIMEN.ID.S6` is linked to  0 `GENOMIC_FILE.ID`  |
| `BIOSPECIMEN.ID.S8` is linked to  0 `GENOMIC_FILE.ID`  |
| `BIOSPECIMEN.ID.S11` is linked to  0 `GENOMIC_FILE.ID` |
| `BIOSPECIMEN.ID.S9` is linked to  0 `GENOMIC_FILE.ID`  |
| `BIOSPECIMEN.ID.S10` is linked to  0 `GENOMIC_FILE.ID` |
| `BIOSPECIMEN.ID.S3` is linked to  0 `GENOMIC_FILE.ID`  |
| `BIOSPECIMEN.ID.S1` is linked to  0 `GENOMIC_FILE.ID`  |
| `BIOSPECIMEN.ID.S5` is linked to  0 `GENOMIC_FILE.ID`  |

| Location   | Values                                               |
|:-----------|:-----------------------------------------------------|
| sfp2.csv   | `NA`,`S2`,`S7`                                       |
| sg.csv     | `S10`,``S7``                                         |
| sp.csv     | `S11`,``S10``,```S7```                               |
| spf.csv    | `S1`,`S3`,`S4`,`S5`,`S6`,`S8`,`S9`,``S2``,````S7```` |

#### ❌ Each BIOSPECIMEN.ID links to exactly 1 BIOSPECIMEN.ANALYTE

| Errors                                                     |
|:-----------------------------------------------------------|
| `BIOSPECIMEN.ID.NA` is linked to  0 `BIOSPECIMEN.ANALYTE`  |
| `BIOSPECIMEN.ID.S7` is linked to  0 `BIOSPECIMEN.ANALYTE`  |
| `BIOSPECIMEN.ID.S2` is linked to  0 `BIOSPECIMEN.ANALYTE`  |
| `BIOSPECIMEN.ID.S4` is linked to  0 `BIOSPECIMEN.ANALYTE`  |
| `BIOSPECIMEN.ID.S6` is linked to  0 `BIOSPECIMEN.ANALYTE`  |
| `BIOSPECIMEN.ID.S8` is linked to  0 `BIOSPECIMEN.ANALYTE`  |
| `BIOSPECIMEN.ID.S11` is linked to  0 `BIOSPECIMEN.ANALYTE` |
| `BIOSPECIMEN.ID.S9` is linked to  0 `BIOSPECIMEN.ANALYTE`  |
| `BIOSPECIMEN.ID.S10` is linked to  0 `BIOSPECIMEN.ANALYTE` |
| `BIOSPECIMEN.ID.S3` is linked to  0 `BIOSPECIMEN.ANALYTE`  |
| `BIOSPECIMEN.ID.S1` is linked to  0 `BIOSPECIMEN.ANALYTE`  |
| `BIOSPECIMEN.ID.S5` is linked to  0 `BIOSPECIMEN.ANALYTE`  |

| Location   | Values                                               |
|:-----------|:-----------------------------------------------------|
| sfp2.csv   | `NA`,`S2`,`S7`                                       |
| sg.csv     | `S10`,``S7``                                         |
| sp.csv     | `S11`,``S10``,```S7```                               |
| spf.csv    | `S1`,`S3`,`S4`,`S5`,`S6`,`S8`,`S9`,``S2``,````S7```` |

#### ❌ Each GENOMIC_FILE.URL_LIST links to exactly 1 GENOMIC_FILE.ID

| Errors                                                                    |
|:--------------------------------------------------------------------------|
| `GENOMIC_FILE.URL_LIST.['foo/s10.txt']` is linked to  0 `GENOMIC_FILE.ID` |
| `GENOMIC_FILE.URL_LIST.['foo/s5.txt']` is linked to  0 `GENOMIC_FILE.ID`  |
| `GENOMIC_FILE.URL_LIST.['foo/s7.txt']` is linked to  0 `GENOMIC_FILE.ID`  |
| `GENOMIC_FILE.URL_LIST.['foo/s11.txt']` is linked to  0 `GENOMIC_FILE.ID` |
| `GENOMIC_FILE.URL_LIST.['foo/s9.txt']` is linked to  0 `GENOMIC_FILE.ID`  |

| Location   | Values                                              |
|:-----------|:----------------------------------------------------|
| pg.csv     | `['foo/s11.txt']`,`['foo/s5.txt']`,`['foo/s9.txt']` |
| sg.csv     | `['foo/s10.txt']`,`['foo/s7.txt']`                  |

#### ✅ Each BIOSPECIMEN.ID links to at most 1 BIOSPECIMEN.SAMPLE_PROCUREMENT

🔘 Each GENOMIC_FILE.ID links to exactly 1 SEQUENCING.ID

🔘 Each SEQUENCING.ID links to at least 1 GENOMIC_FILE.ID

🔘 Each GENOMIC_FILE.ID links to exactly 1 BIOSPECIMEN.ID

🔘 Each PARTICIPANT.GENDER links to at least 1 PARTICIPANT.ID

🔘 Each PARTICIPANT.ETHNICITY links to at least 1 PARTICIPANT.ID

🔘 Each PARTICIPANT.RACE links to at least 1 PARTICIPANT.ID

🔘 Each PARTICIPANT.SPECIES links to at least 1 PARTICIPANT.ID

🔘 Each PARTICIPANT.ENROLLMENT_AGE_DAYS links to at least 1 PARTICIPANT.ID

🔘 Each BIOSPECIMEN.ANALYTE links to at least 1 BIOSPECIMEN.ID

🔘 Each BIOSPECIMEN.COMPOSITION links to at least 1 BIOSPECIMEN.ID

🔘 Each BIOSPECIMEN.TISSUE_TYPE links to at least 1 BIOSPECIMEN.ID

🔘 Each BIOSPECIMEN.ANATOMY_SITE links to at least 1 BIOSPECIMEN.ID

🔘 Each BIOSPECIMEN.TUMOR_DESCRIPTOR links to at least 1 BIOSPECIMEN.ID

🔘 Each BIOSPECIMEN.EVENT_AGE_DAYS links to at least 1 BIOSPECIMEN.ID

🔘 Each BIOSPECIMEN.SPATIAL_DESCRIPTOR links to at least 1 BIOSPECIMEN.ID

🔘 Each BIOSPECIMEN.SHIPMENT_ORIGIN links to at least 1 BIOSPECIMEN.ID

🔘 Each BIOSPECIMEN.SHIPMENT_DATE links to at least 1 BIOSPECIMEN.ID

🔘 Each BIOSPECIMEN.CONCENTRATION_MG_PER_ML links to at least 1 BIOSPECIMEN.ID

🔘 Each BIOSPECIMEN.VOLUME_UL links to at least 1 BIOSPECIMEN.ID

🔘 Each BIOSPECIMEN.SAMPLE_PROCUREMENT links to at least 1 BIOSPECIMEN.ID

🔘 Each GENOMIC_FILE.HARMONIZED links to at least 1 GENOMIC_FILE.ID

🔘 Each GENOMIC_FILE.ID links to exactly 1 GENOMIC_FILE.HARMONIZED

🔘 Each GENOMIC_FILE.REFERENCE_GENOME links to at least 1 GENOMIC_FILE.ID

🔘 Each GENOMIC_FILE.ID links to exactly 1 GENOMIC_FILE.REFERENCE_GENOME

🔘 Each GENOMIC_FILE.ID links to exactly 1 GENOMIC_FILE.URL_LIST

🔘 Each SEQUENCING.LIBRARY_NAME links to exactly 1 SEQUENCING.ID

🔘 Each SEQUENCING.ID links to at most 1 SEQUENCING.LIBRARY_NAME

🔘 Each SEQUENCING.STRATEGY links to at least 1 SEQUENCING.ID

🔘 Each SEQUENCING.ID links to exactly 1 SEQUENCING.STRATEGY

🔘 Each SEQUENCING.PAIRED_END links to at least 1 SEQUENCING.ID

🔘 Each SEQUENCING.ID links to exactly 1 SEQUENCING.PAIRED_END

🔘 Each SEQUENCING.PLATFORM links to at least 1 SEQUENCING.ID

🔘 Each SEQUENCING.ID links to at most 1 SEQUENCING.PLATFORM

🔘 Each SEQUENCING.INSTRUMENT links to at least 1 SEQUENCING.ID

🔘 Each SEQUENCING.ID links to at most 1 SEQUENCING.INSTRUMENT

#### ❌ All resolved links are hierarchically direct

| Errors                                                                         |
|:-------------------------------------------------------------------------------|
| `GENOMIC_FILE.URL_LIST.['foo/s11.txt']` is linked to  ['`BIOSPECIMEN.ID.S11`'] |
| `GENOMIC_FILE.URL_LIST.['foo/s10.txt']` is linked to  ['`BIOSPECIMEN.ID.S10`'] |
| `GENOMIC_FILE.URL_LIST.['foo/s5.txt']` is linked to  ['`BIOSPECIMEN.ID.S5`']   |
| `GENOMIC_FILE.URL_LIST.['foo/s7.txt']` is linked to  ['`BIOSPECIMEN.ID.S7`']   |
| `GENOMIC_FILE.URL_LIST.['foo/s9.txt']` is linked to  ['`BIOSPECIMEN.ID.S9`']   |

| Location   | Values                                                                                                                                   |
|:-----------|:-----------------------------------------------------------------------------------------------------------------------------------------|
| pg.csv     | `GENOMIC_FILE.URL_LIST.['foo/s11.txt']`,`GENOMIC_FILE.URL_LIST.['foo/s5.txt']`,`GENOMIC_FILE.URL_LIST.['foo/s9.txt']`                    |
| sfp2.csv   | `BIOSPECIMEN.ID.S7`                                                                                                                      |
| sg.csv     | `BIOSPECIMEN.ID.S10`,`BIOSPECIMEN.ID.`BIOSPECIMEN.ID.S7``,`GENOMIC_FILE.URL_LIST.['foo/s10.txt']`,`GENOMIC_FILE.URL_LIST.['foo/s7.txt']` |
| sp.csv     | `BIOSPECIMEN.ID.S11`,`BIOSPECIMEN.ID.`BIOSPECIMEN.ID.S10``,`BIOSPECIMEN.ID.`BIOSPECIMEN.ID.`BIOSPECIMEN.ID.S7```                         |
| spf.csv    | `BIOSPECIMEN.ID.S5`,`BIOSPECIMEN.ID.S9`,`BIOSPECIMEN.ID.`BIOSPECIMEN.ID.`BIOSPECIMEN.ID.`BIOSPECIMEN.ID.S7````                           |

### `Fail: 0` `Success: 0` `Did Not Run: 36`

🔘 PARTICIPANT.RACE must be one of ['American Indian or Alaska Native', 'Asian', 'Black or African American', 'More Than One Race', 'Native Hawaiian or Other Pacific Islander', 'Other', 'White'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 PARTICIPANT.ENROLLMENT_AGE_DAYS must be a number x such that 0 <= x <= 32872.

🔘 PARTICIPANT.ETHNICITY must be one of ['Hispanic or Latino', 'Not Hispanic or Latino'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 PARTICIPANT.GENDER must be one of ['Female', 'Male', 'Other'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 PARTICIPANT.SPECIES must be one of ['Canis lupus familiaris', 'Homo Sapiens'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 PARTICIPANT.CONSENT_TYPE must be one of ['DS-CHD', 'DS-CHD-IRB', 'DS-OBD-MDS', 'DS-OBDR-MDS', 'DS-OC-PUB-MDS', 'GRU', 'HMB-IRB', 'HMB-MDS'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 BIOSPECIMEN.EVENT_AGE_DAYS must be a number x such that 0 <= x <= 32872.

🔘 BIOSPECIMEN.ANALYTE must be one of ['DNA', 'Other', 'RNA', 'Virtual'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 BIOSPECIMEN.CONCENTRATION_MG_PER_ML must be a number > 0.

🔘 BIOSPECIMEN.SAMPLE_PROCUREMENT must be one of ['Autopsy', 'Biopsy', 'Blood Draw', 'Gross Total Resection', 'Other', 'Subtotal Resection'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 BIOSPECIMEN.SHIPMENT_DATE must be a valid representation of a Date or Datetime object.

🔘 BIOSPECIMEN.VOLUME_UL must be a number > 0.

🔘 DIAGNOSIS.EVENT_AGE_DAYS must be a number > 0.

🔘 DIAGNOSIS.CATEGORY must be one of ['Cancer', 'Other', 'Structural Birth Defect'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 GENOMIC_FILE.AVAILABILITY must be one of ['Cold Storage', 'Immediate Download'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 GENOMIC_FILE.DATA_TYPE must be one of ['Aligned Reads', 'Aligned Reads Index', 'Expression', 'Gene Expression', 'Gene Fusions', 'Histology Images', 'Isoform Expression', 'Operation Reports', 'Other', 'Pathology Reports', 'Radiology Images', 'Radiology Reports', 'Simple Nucleotide Variations', 'Somatic Copy Number Variations', 'Somatic Structural Variations', 'Unaligned Reads', 'Variant Calls', 'Variant Calls Index', 'gVCF', 'gVCF Index'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 GENOMIC_FILE.SIZE must be a number >= 0.

🔘 OUTCOME.EVENT_AGE_DAYS must be a number x such that 0 <= x <= 32872.

🔘 OUTCOME.DISEASE_RELATED must be one of ['No', 'Yes'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 OUTCOME.VITAL_STATUS must be one of ['Alive', 'Deceased'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 PHENOTYPE.EVENT_AGE_DAYS must be a number x such that 0 <= x <= 32872.

🔘 PHENOTYPE.OBSERVED must be one of ['Negative', 'Positive'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 READ_GROUP.PAIRED_END must be one of ['1', '2'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 READ_GROUP.LANE_NUMBER must be a number > 0.

🔘 READ_GROUP.QUALITY_SCALE must be one of ['Illumina13', 'Illumina15', 'Illumina18', 'Sanger', 'Solexa'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 SEQUENCING.DATE must be a valid representation of a Date or Datetime object.

🔘 SEQUENCING.STRATEGY must be one of ['Linked-Read WGS (10x Chromium)', 'Other', 'RNA-Seq', 'Targeted Sequencing', 'WGS', 'WXS', 'miRNA-Seq'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 SEQUENCING.LIBRARY_STRAND must be one of ['First Stranded', 'Other', 'Second Stranded', 'Unstranded'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 SEQUENCING.MAX_INSERT_SIZE must be a number > 0.

🔘 SEQUENCING.MEAN_DEPTH must be a number > 0.

🔘 SEQUENCING.MEAN_INSERT_SIZE must be a number > 0.

🔘 SEQUENCING.MEAN_READ_LENGTH must be a number > 0.

🔘 SEQUENCING.PAIRED_END must be one of ['False', 'True'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 SEQUENCING.PLATFORM must be one of ['Complete Genomics', 'Illumina', 'Ion Torrent', 'LS454', 'Other', 'PacBio', 'SOLiD'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

🔘 SEQUENCING.TOTAL_READS must be a number > 0.

🔘 STUDY_FILE.AVAILABILITY must be one of ['Cold Storage', 'Immediate Download'] or ['Not Allowed To Collect', 'Not Applicable', 'Not Available', 'Not Reported', 'Reported Unknown']

### `Fail: 0` `Success: 0` `Did Not Run: 0`

### CODE REVIEW

File: kf_lib_data_ingest/validation/reporting/base.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+SAMPLE_VALIDATION_RESULTS = os.path.abspath("./sample_validation_results.py")
+REQUIRED_RESULT_KEYS = {"type", "description", "is_applicable", "errors"}
+PASSED = "success"
+FAILED = "fail"
+NA = "did not run"
+RESULT_TO_EMOJI = {PASSED: "✅", FAILED: "❌", NA: "🔘"}
+
+
+class AbstractReportBuilder(ABC):
+    def __init__(self, output_dir=None, setup_logger=False):
```
Comment (gsantia) [inline kf_lib_data_ingest/validation/reporting/base.py:23]: Considering how large this class is, maybe it'd help to have at least a short blurb on what it's used for. It being an abc gives me some idea but I'm not totally sure without going through line by line 😁

File: kf_lib_data_ingest/validation/reporting/base.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+    def build(self, results, **report_kwargs):
+        """
+        Validate results schema
+        Build validation report content from validation result dicts
+        Write validation report files to disk
+
+        :param results: See _build
+        :type results: See _build
+        :param report_kwargs: Keyword args passed to subclasses _build
```
Comment (gsantia) [inline kf_lib_data_ingest/validation/reporting/base.py:88]:
```suggestion
        :param report_kwargs: Keyword args passed to subclass _build implementation
```

File: kf_lib_data_ingest/validation/reporting/base.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        :type results: list of dicts
+        """
+        assert_safe_type(results["type_counts"], dict)
+        assert_safe_type(results["files_validated"], list)
+        assert_safe_type(results["validation"], list)
+
+        for r in results["validation"]:
+            # Check for required keys
+            diff = REQUIRED_RESULT_KEYS - set(r.keys())
+            if diff:
```
Comment (gsantia) [inline kf_lib_data_ingest/validation/reporting/base.py:143]:
```suggestion
            if REQUIRED_RESULT_KEYS.difference(r.keys()):
```

File: kf_lib_data_ingest/validation/reporting/base.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+            try:
+                errs = r["errors"]
+                if r["type"] == "relationship":
+                    for e in errs:
+                        t, v = e["from"]
+                        if e["to"]:
+                            for to_node in e["to"]:
+                                t1, v1 = to_node
+                        for (t2, v2), files in e["locations"].items():
+                            pass
```
Comment (gsantia) [inline kf_lib_data_ingest/validation/reporting/base.py:166]: What's going on here?

Comment (znatty22) [reply kf_lib_data_ingest/validation/reporting/base.py:188]: yes

File: kf_lib_data_ingest/validation/reporting/base.py [hunk 5]
```diff
@@ -20,6 +20,21 @@
...
+    """
+    Abstract base class defines the required interface for all
+    subclass validation report builders
+
+    Subclasses must implement methods:
+    - `_build`: builds the report content in a particular format
+    - `_write_report`: writes the report content to disk.
+
+    The public `build` method will call `_validate_result_schema` to
+    ensure validation results adhere to the expected the structure/format
```
Comment (gsantia) [inline kf_lib_data_ingest/validation/reporting/base.py:31]:
```suggestion
    ensure validation results adhere to the expected structure/format
```

Comment (gsantia) [reply kf_lib_data_ingest/validation/reporting/base.py:31]: This is my only comment regarding the code. For the overall structure, it looks good to me. But maybe @fiendish should review as well.

File: kf_lib_data_ingest/validation/reporting/markdown.py [hunk 6]
```diff
@@ -40,6 +40,10 @@ def _build(self, results, title=DEFAULT_REPORT_TITLE):
         :param title: Report title
         :type title: str
         """
+        self.counts = results['counts']
+        self.files_validated = results['files_validated']
+        self.results = results['validation']
```
Comment (gsantia) [inline kf_lib_data_ingest/validation/reporting/markdown.py:45]:
```suggestion
        self.counts = results["counts"]
        self.files_validated = results["files_validated"]
        self.results = results["validation"]
```

File: kf_lib_data_ingest/validation/reporting/markdown.py [hunk 7]
```diff
@@ -104,6 +108,9 @@ def _tests_section_md(self, results, test_type):
             if r["type"] != test_type:
                 continue
             result_code = self._result_code(r)
+            self.logger.debug(
+                f"Building results for {r['description']}"
```
Comment (gsantia) [inline kf_lib_data_ingest/validation/reporting/markdown.py:112]:
```suggestion
                f"Building results for {r["description"]}"
```

File: kf_lib_data_ingest/validation/reporting/markdown.py [hunk 8]
```diff
@@ -155,32 +165,69 @@ def _result_to_md(self, rd, test_type):
...
+            locs = defaultdict(set)
+
+            # -- Errors --
+
+            # Rollup errors into 1 line when all from_nodes are linked to
+            # 0 to_nodes (e.g. All PARTICIPANT.ID are linked to [])
+            err_rollup = False
+            from_type, _ = rd['errors'][0]['from']
+            if (
+                len(rd['errors']) == self.counts[from_type] and
```
Comment (gsantia) [inline kf_lib_data_ingest/validation/reporting/markdown.py:179]:
```suggestion
            from_type, _ = rd["errors"][0]["from"]
            if (
                len(rd"errors"]) == self.counts[from_type] and
```

File: kf_lib_data_ingest/validation/reporting/markdown.py [hunk 9]
```diff
@@ -155,32 +165,69 @@ def _result_to_md(self, rd, test_type):
...
+                all([not e['to'] for e in errors])
```
Comment (gsantia) [inline kf_lib_data_ingest/validation/reporting/markdown.py:180]:
```suggestion
                all([not e["to"] for e in errors])
```

File: kf_lib_data_ingest/validation/reporting/markdown.py [hunk 10]
```diff
@@ -155,32 +165,69 @@ def _result_to_md(self, rd, test_type):
...
+                errors.append(
+                    {
+                        "Errors": f"All `{from_type.replace('|', REPLACE_PIPE)}` "
+                        "are linked to []"
+                    }
+                )
+            # Create formatted errors
+            else:
+                for e in rd["errors"]:
+                    prefix = f"{tuple_to_str(e['from'])} is linked to "
```
Comment (gsantia) [inline kf_lib_data_ingest/validation/reporting/markdown.py:192]:
```suggestion
                    prefix = f"{tuple_to_str(e["from"])} is linked to "
```

File: kf_lib_data_ingest/validation/reporting/markdown.py [hunk 11]
```diff
@@ -155,32 +165,69 @@ def _result_to_md(self, rd, test_type):
...
+                    if e["to"]:
+                        suffix = [f"`{to_node[1]}`" for to_node in e["to"]]
+                    else:
+                        suffix = e["to"]
+                    err_str = f"{prefix} {suffix}".replace("|", REPLACE_PIPE)
+                    errors.append({"Errors": err_str})
+
+            # -- File locations --
+            for e in rd['errors']:
```
Comment (gsantia) [inline kf_lib_data_ingest/validation/reporting/markdown.py:201]:
```suggestion
            for e in rd["errors"]:
```

File: kf_lib_data_ingest/validation/reporting/markdown.py [hunk 12]
```diff
@@ -155,32 +165,69 @@ def _result_to_md(self, rd, test_type):
...
+                for (typ, val), files in e["locations"].items():
+                    for f in files:
+                        locs[f].add(f'`{val}`')
+
+            # Rollup locations - if number of error values in a file ==
+            # number of total errors for the test, then don't list out every
+            # error value in that file
+            loc_rows = []
+            for loc, vals in locs.items():
+                if len(vals) == len(rd['errors']):
```
Comment (gsantia) [inline kf_lib_data_ingest/validation/reporting/markdown.py:211]:
```suggestion
                if len(vals) == len(rd["errors"]):
```

File: kf_lib_data_ingest/validation/reporting/markdown.py [hunk 13]
```diff
@@ -155,32 +165,69 @@ def _result_to_md(self, rd, test_type):
...
                 pandas.DataFrame(errors).to_markdown(index=False)
             )
-            test_markdown.append("")
-            test_markdown.append(
-                pandas.DataFrame(locs).to_markdown(index=False)
-            )
+            if not err_rollup:
+                test_markdown.append("")
+                test_markdown.append(
+                    pandas.DataFrame(loc_rows).sort_values(by='Location')
```
Comment (gsantia) [inline kf_lib_data_ingest/validation/reporting/markdown.py:228]:
```suggestion
                    pandas.DataFrame(loc_rows).sort_values(by="Location")
```

**Comment by fiendish** [comment]
Was the markdown report going to get replaced by a full HTML one with collapsible sections?

**Review by gsantia — APPROVED** [review]
👍

**Comment by znatty22** [comment]
> Was the markdown report going to get replaced by one with collapsible sections?

I didn't realize you could do collapsible sections in markdown. Will incorporate that and ask for another review!

File: kf_lib_data_ingest/validation/reporting/base.py [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+                    assert_all_safe_type(errs.values(), int)
+
+            except Exception:
+                self.logger.error(
+                    "Badly formatted `errors` in validation result dict: "
+                    f"\n{pformat(r)}\n"
+                    f"Please see {SAMPLE_VALIDATION_RESULTS} "
+                    f"for expected validation results format."
+                )
+                raise
```
Comment (fiendish) [inline kf_lib_data_ingest/validation/reporting/base.py:225]: feels weird to have this here and not just in a unit test

**Review by fiendish — APPROVED** [review]

File: kf_lib_data_ingest/validation/reporting/base.py (hunk 14 above)
Comment (znatty22) [reply kf_lib_data_ingest/validation/reporting/base.py:225]: Yea I guess I could remove it now that the interface is stable 🤷‍♀️. Plus people can look at `sample_validation_results.py` for a look at the result format

**Review by fiendish — APPROVED** [review]
📝

**Pull request merged**